lib/friendly_shipping/services/ship_engine/parse_address_validation_response.rb

```diff
@@ -0,0 +1,71 @@
+# frozen_string_literal: true
+
+require 'json'
+
+module FriendlyShipping
+  module Services
+    class ShipEngine
+      class ParseAddressValidationResponse
+        extend Dry::Monads::Result::Mixin
+
+        class << self
+          # @param [FriendlyShipping::Request] request
+          # @param [FriendlyShipping::Response] response
+          # @return [Success<ApiResult>, Failure<ApiFailure>]
+          def call(request:, response:)
+            parsed_json = JSON.parse(response.body)
+            errors = parsed_json.map do |result|
+              result['messages'].map { |message| message['message'] }.join(", ") if result['status'] == "error"
+            end.compact
+            if errors.any?
+              Failure(
+                ApiFailure.new(
+                  errors,
+                  original_request: request,
+                  original_response: response
+                )
+              )
+            else
+              locations = parsed_json.map do |result|
+                address = result['matched_address']
+                Physical::Location.new(
+                  name: address['name'],
+                  email: address['email'],
+                  phone: address['phone'],
+                  company_name: address['company_name'],
+                  address1: address['address_line1'],
+                  address2: address['address_line2'],
+                  address3: address['address_line3'],
+                  city: address['city_locality'],
+                  zip: address['postal_code'],
+                  region: address['state_province'],
+                  country: address['country_code'],
+                  address_type: address_type(address)
+                )
+              end
+              Success(
+                ApiResult.new(
+                  locations,
+                  original_request: request,
+                  original_response: response
+                )
+              )
+            end
+          end
```

Comment:
What do you think about adding `rescue JSON::ParserError` here? I can't remember how ShipEngine returns errors when they are down or having issues.

Reply:
That's a good idea.